src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -394,6 +399,228 @@ begin
+is_trail.of_append_right (by rwa ← take_spec _ h at hc)
+
+lemma is_path.take_until {u v w : V} {p : G.walk v w} (hc : p.is_path) (h : u ∈ p.support) :
+  (p.take_until u h).is_path :=
+is_path.of_append_left (by rwa ← take_spec _ h at hc)
+
+lemma is_path.drop_until {u v w : V} (p : G.walk v w) (hc : p.is_path) (h : u ∈ p.support) :
+  (p.drop_until u h).is_path :=
+is_path.of_append_right (by rwa ← take_spec _ h at hc)
+
+/-- Rotate a loop walk such that it is centered at the given vertex. -/
+def rotate {u v : V} (c : G.walk v v) (h : u ∈ c.support) : G.walk u u :=
+(c.drop_until u h).append (c.take_until u h)
+
+@[simp]
+lemma rotate_support {u v : V} (c : G.walk v v) (h : u ∈ c.support) :
+  (c.rotate h).support.tail ~r c.support.tail :=
+begin
+  simp only [rotate, tail_support_append],
+  apply list.is_rotated.trans list.is_rotated_append,
+  rw [←tail_support_append, take_spec],
+end
+
+lemma rotate_edges {u v : V} (c : G.walk v v) (h : u ∈ c.support) :
+  (c.rotate h).edges ~r c.edges :=
+begin
+  simp only [rotate, edges_append],
+  apply list.is_rotated.trans list.is_rotated_append,
+  rw [←edges_append, take_spec],
+end
+
+lemma is_trail.rotate {u v : V} {c : G.walk v v} (hc : c.is_trail) (h : u ∈ c.support) :
+  (c.rotate h).is_trail :=
+begin
+  rw [is_trail_def, (c.rotate_edges h).perm.nodup_iff],
+  exact hc.edges_nodup,
+end
+
+lemma is_circuit.rotate {u v : V} {c : G.walk v v} (hc : c.is_circuit) (h : u ∈ c.support) :
+  (c.rotate h).is_circuit :=
+begin
+  refine ⟨hc.to_trail.rotate _, _⟩,
+  cases c,
+  { exact (hc.ne_nil rfl).elim, },
+  { intro hn,
+    have hn' := congr_arg length hn,
+    rw [rotate, length_append, add_comm, ← length_append, take_spec] at hn',
+    simpa using hn', },
+end
+
+lemma is_cycle.rotate {u v : V} {c : G.walk v v} (hc : c.is_cycle) (h : u ∈ c.support) :
+  (c.rotate h).is_cycle :=
+begin
+  refine ⟨hc.to_circuit.rotate _, _⟩,
+  rw list.is_rotated.nodup_iff (rotate_support _ _),
+  exact hc.support_nodup,
+end
+
+end walk_decomp
+
+end walk
+
+/-! ### Walks to paths -/
+
+/-- The type for paths between two vertices. -/
+abbreviation path (u v : V) := {p : G.walk u v // p.is_path}
+
+namespace walk
+variables {G} [decidable_eq V]
+
+/-- Given a walk, produces a walk with the same endpoints and no repeated vertices. -/
+def to_path_aux : Π {u v : V}, G.walk u v → G.walk u v
+| u v nil := nil
+| u v (cons ha p) :=
+  let p' := p.to_path_aux
+  in if hs : u ∈ p'.support
+     then p'.drop_until u hs
+     else cons ha p'
+
+lemma to_path_aux_is_path {u v : V} (p : G.walk u v) : p.to_path_aux.is_path :=
+begin
+  induction p,
+  { simp!, },
+  { simp only [to_path_aux],
+    split_ifs,
+    { apply is_path.drop_until,
+      assumption, },
+    { simp [*, cons_is_path_iff], } },
+end
+
+/-- Given a walk, produces a path with the same endpoints using `simple_graph.walk.to_path_aux`. -/
+def to_path {u v : V} (p : G.walk u v) : G.path u v := ⟨p.to_path_aux, p.to_path_aux_is_path⟩
```

Comment:
Do you have a lemma stating the minimality of `to_path`?

Reply:
I've added a lemma that the length doesn't get longer. (This function doesn't guarantee a path of minimal length.)

Reply:
But it does guarantee that any path containing it is actually itself.

Reply:
I'm happy with the future PR solution 😄